File: packages/sdk-browser/src/dsn.ts
import type { ParsedDsn } from './types.js';

// User-facing DSN format (matches what the api emits in the project-create response and shows in
// the web UI's "Copy DSN" modal): arguslog://<publicKey>@<host>/api/<projectId>
//
// The custom `arguslog://` scheme keeps the DSN string visually distinct from the actual transport
// URL and prevents users from copy-pasting it into a browser hoping it'll open something. The SDK
// derives the real transport scheme from the host: localhost / loopback / RFC1918 → http (dev),
// everything else → https.
const DSN_RE = /^arguslog:\/\/([^@]+)@([^/]+)\/api\/([^/?#]+)$/;

export class InvalidDsnError extends Error {
  constructor(dsn: string) {
    super(`Invalid Arguslog DSN: ${dsn}`);
    this.name = 'InvalidDsnError';
  }
}

export function parseDsn(dsn: string): ParsedDsn {
  const match = DSN_RE.exec(dsn);
  if (!match) {
    throw new InvalidDsnError(dsn);
  }
  const [, publicKey, host, projectId] = match;
  if (!publicKey || !host || !projectId) {
    throw new InvalidDsnError(dsn);
  }
  const protocol: 'http' | 'https' = isLoopback(host) ? 'http' : 'https';
  return {
    protocol,
    publicKey,
    host,
    projectId,
    ingestUrl: `${protocol}://${host}/api/${projectId}/events`,
  };
}

function isLoopback(host: string): boolean {
  // IPv6 literals come bracketed (e.g. [::1]:8080); strip the [..] part. IPv4/hostnames have a
  // single optional :port suffix.
  const bare = host.startsWith('[')
    ? host.slice(0, host.indexOf(']') + 1)
    : (host.split(':')[0] ?? host);
  return (
    bare === 'localhost' ||
    bare.startsWith('127.') ||
    bare === '0.0.0.0' ||
    bare === '[::1]' ||
    bare === '::1'
  );
}

File: java-sdk/src/main/java/org/arguslog/sdk/Dsn.java
package org.arguslog.sdk;

import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public record Dsn(
    String publicKey, String host, String scheme, String projectId, String ingestUrl) {

  // User-facing DSN format (matches what the api emits in the project-create response and shows
  // in the web UI's "Copy DSN" modal): arguslog://<publicKey>@<host>/api/<projectId>
  // Kept in lockstep with packages/sdk-browser/src/dsn.ts.
  private static final Pattern DSN_RE =
      Pattern.compile("^arguslog://([^@]+)@([^/]+)/api/([^/?#]+)$");

  public static Dsn parse(String raw) {
    Objects.requireNonNull(raw, "dsn");
    Matcher m = DSN_RE.matcher(raw);
    if (!m.matches()) {
      throw new IllegalArgumentException("Invalid DSN: " + raw);
    }
    String publicKey = m.group(1);
    String host = m.group(2);
    String projectId = m.group(3);
    if (publicKey.isEmpty() || host.isEmpty() || projectId.isEmpty()) {
      throw new IllegalArgumentException("Invalid DSN: " + raw);
    }
    String scheme = isLoopback(host) ? "http" : "https";
    String ingestUrl = scheme + "://" + host + "/api/" + projectId + "/events";
    return new Dsn(publicKey, host, scheme, projectId, ingestUrl);
  }

  private static boolean isLoopback(String host) {
    String bare =
        host.startsWith("[")
            ? host.substring(0, host.indexOf(']') + 1)
            : host.split(":", 2)[0];
    return bare.equals("localhost")
        || bare.startsWith("127.")
        || bare.equals("0.0.0.0")
        || bare.equals("[::1]")
        || bare.equals("::1");
  }
}
